feat: 104269: add gammaScale function to tlisp

FlufflesTheMicrosaur opened the pull request:
https://ministry.kronosaur.com/record.hexm?id=104269

gcabbage commented on Alchemy/CodeChain/DefPrimitives.h:
```diff
@@ -186,6 +186,13 @@ static PRIMITIVEPROCDEF g_DefPrimitives[] =
 			NULL,	0,	},
 
+		{	"gammaScale",		fnMathNumerals,	FN_MATH_GAMMA_SCALE_NUMERALS,
+			"(gammaScale var inMin inMax outMin outMax [gamma=1.0]) -> int or double of gammascaled value relative to input and output range.\n"
+			"Returns int if both outMin and outMax are ints, otherwise returns a double. Gamma specifies an exponent relationship curve between input and output ranges.\n"
```
Most multi line help messages have a blank line between the signature and the extended text which helps with readability in the console.

```suggestion
			"(gammaScale var inMin inMax outMin outMax [gamma=1.0]) -> scaled value (int or real)\n\n"
			
			"Scales var from range inMin to inMax to the output scale outMin to outMax.\n"
			"Returns int if both outMin and outMax are ints, otherwise returns a real. Gamma specifies an exponent relationship curve between input and output ranges.\n"
```

gcabbage reviewed (commented):
Can you call the new function `mathScale` and delete the old lambda from CoreCode.xml?

That should be fully compatible between versions and would not require any replacement of math->gamma calls e.g. #166. This would also immediately benefit any mods which make use of mathScale.

I have checked with a save firing a smartcannon and can swap back and forth between a build from master and a modified version of this branch with the native function named as `mathScale` (and the lambda version removed)

Has `mathScaleReal` been in any release yet? If not I'd be tempted to remove that too so only this new native version will be used (or if necessary define it as `(setq mathScaleReal mathScale)`)

FlufflesTheMicrosaur commented:
Ok I did a bunch of testing (its posted on the discord, and I put the specific branch I'm testing with on my github: preview/2.0api55preview1 - built w/ Preview For Contributors config)

Using an AMD R9 7950X, I got the following perf figures: ("us" is supposed to be microseconds)
uncached: (running it every few seconds, since a bunch of other stuff is happening in the game)
~40-80us for mathScale/mathScaleReal (this is normally what the game will experience)
~4-5us for gammaScale (including w/ aliases - doesnt make a difference, which makes sense given the codepath)
~10-20us for the "maximum compatibility lambdas" above
highly cached: (running it in a loop 1000000 times with the same args every time)
~6-7us for mathScale/mathScaleReal
~430ns for gammaScale & aliases
~1.8us for the "maximum compatibility lambdas" above

Ultimately switching is a big win in terms of performance (effect heavy scenes/ships were performing noticeably better, like the TX2 Italia, ORE** Odin, and EUC* Yavuz and painting became the new bottleneck for them when firing their weapons), but the choice of more or less compatibility options is ultimately up to George - we will raise it on monday w/ him. Switching the core codebase to natively use gammaScale is easy enough - there is already a PR for this and it can be done with a trivial find and replace. Same goes for any expansions (CC/EP/TSB/etc). For unmaintained mods, using the compatibility lambdas would still be a big help performance-wise. 

*Can be seen via TSB 
**Not yet released

gcabbage commented:
Definitely worthwhile for the speed increase. Given the way mathScale is currently used my initial feeling is that the increased speed from the aliases is more useful than the "maximum compatibility" lambda. But as you say it's up to George / someone using the functions in a mod.

FlufflesTheMicrosaur commented:
Oh! yeah ok thats definitely a bug, good catch.

I was just copying over george's original math without looking too hard at it and then adapted it because of the premultiplication issues, and then NMS noticing that george didnt even handle the inverted input range situation.

FlufflesTheMicrosaur commented:
Ok I fixed the math, renamed vars and reorganized the code (With some explanations) to hopefully make it actually comprehensible to anyone looking at it in the future.
